### DISCUSSION TITLE
CTT Security Username Password wrong error code

### DISCUSSION THREAD

**Discussion by johannwesely** [eclipse-milo/milo#1324] [General]
Hi in Security User Name Password there seems to bee also an wrong error code:
![image](https://github.com/user-attachments/assets/dba200a3-918f-4ebc-8ca9-ae440b940897)
i changed Line 132 in AbstractIdentityValidator to` throw new UaException(StatusCodes.Bad_IdentityTokenRejected);`
![image](https://github.com/user-attachments/assets/4c0c881f-0b12-4eb0-b082-18524eb35243)
but for case 012 it seems that the BadUserAccessDenied is overwritten by the new Bad_IdentityTokenrejected. But maybe im wrong? when i debug this i reach in AbstractUsernameIdentityValidator Line 147  where it throws BadUserAccessDenied in Method authenticateUsernameOrThrow but right after this i get to Line 132 in AbstractIdentityValidator. 
Greetings Johann

**Comment by johannwesely** [comment]
In Session Services Session Change User this seems to be the same issue :
![image](https://github.com/user-attachments/assets/adc6fb38-6a03-4bd3-9466-d8f0e11c77d8)

**Comment by johannwesely** [comment]
Hi Kevin, did you have the Time to check on this ?

**Reply by kevinherron** [reply]
Ah, I forgot about this.

I started on it here: https://github.com/eclipse/milo/compare/dev/1.0...1.0-x509-identity

It passes the CTT tests, but needs some work to not re-use the server's certificate validator because it will end up imposing requirements on the identity certificate that are not necessary.

**Reply by kevinherron** [reply]
I'll try to work on it a little this week.

**Comment by kevinherron** [comment]
@johannwesely not sure if this still matters to you, but the new [demo server](https://github.com/digitalpetri/opc-ua-demo-server) running the latest 1.0.0-SNAPSHOT artifact now passes the "Standard 2017 UA Server" profile, which includes these test cases and others you posted about.

(minus the odd timing issue or CTT test case bug)

**Comment by johannwesely** [comment]
Congratulations i really apprechiate your good work. We are still interested in certifieing our milo server but i get stuck a little doing other things but will come back to certification tasks soon. Greetings Johann
